### ISSUE TITLE
2 * suspicious condition ?

### ISSUE THREAD

**Issue by dcb314** [distcc/distcc#179]
1.

src/clirpc.c:200]: (style) Suspicious condition (assignment + comparison); Clarify expression with parentheses.

  if ((ret = dcc_r_token_int(net_fd, "DOTD", &len) == 0)

Maybe better code

  if (((ret = dcc_r_token_int(net_fd, "DOTD", &len)) == 0)

2.

src/compile.c:610]: (style) Suspicious condition (assignment + comparison); Clarify expression with parentheses.

Source code is

      if ((ret = dcc_cpp_maybe(argv, input_fname, &cpp_fname, &cpp_pid) != 0))

Maybe better code

      if ((ret = dcc_cpp_maybe(argv, input_fname, &cpp_fname, &cpp_pid)) != 0)

**Comment by TafThorne** [comment]
The problem line given in 1. is not symmetric, there must be another associated closing ) somewhere.  That could change what is going on if this is a multi-line statement.  

In the current revision the statement in 1. seems to be on [line 203](https://github.com/distcc/distcc/blob/3aeae5f2a6725c19f67c54c45c2ba4273ccc8030/src/clirpc.c#L203) now: 
```
if ((ret = dcc_r_token_int(net_fd, "DOTD", &len) == 0)
                && deps_fname != NULL) {
```
If is multi-line.  `&&` is lower than `==` but higher than `=` and `!=`.  However the existing braces suggest the proposed change would not significantly alter the behaviour.  http://en.cppreference.com/w/c/language/operator_precedence 

This is change is only going to matter if `ret` might fail to be assigned or be changed externally; this is not the case as ret is internal to the function.  

For [the code reference in 2.](https://github.com/distcc/distcc/blob/3aeae5f2a6725c19f67c54c45c2ba4273ccc8030/src/compile.c#L610) it might be even clearer to make the code use the `else` case as if (0) is implicit a failure while if (!0) is a pass.  I guess in C++ there might be some fun casting or perhaps some strange pointer behaviour where checking for void might need the `== 0` test? 


I started looking into this because of the non-symmetric braces.  If I scratch at something I may as well take a proper look.  The suggestion seems OK to me.  I have not spotted any problems it is likely to cause.

**Comment by dcb314** [comment]
>However the existing braces suggest the proposed change would not significantly alter the >behaviour.

False. Newbie error. The order of the assignment and comparison do matter.

This is explained more fully on page 16 of K & R 2, section 1.5.1. RTFM.

**Comment by TafThorne** [comment]
@dcb314 I was clearly speaking about this specific use case.  Please explain why you have suggested a change that makes my statement "However the existing braces suggest the proposed change would not significantly alter the behaviour" false.

**Comment by dcb314** [comment]
Suggest eschew prolixity for this trivial case. The code is clearly broken, I've suggested a 
reasonable fix.

**Comment by TafThorne** [comment]
@dcb314 I agree with the sentiment but when I looked over the code and said it was OK you then stated that my analysis of the fix being OK was false.  If that is the case the suggested fix is not reasonable as it is not OK by your own statement.  Which is it?  Was I correct in stating that your suggested fix was fine in this example, as when including the line missing in the example no significant alteration of behaviour was evident.  Or did I miss some nuance in the code of this system that made your request for me to read a manual significant despite my having already eluded to the specification to which you reference me?  Newbies cannot learn if questions go unanswered.

**Comment by paranormal** [comment]
if ((ret = dcc_r_token_int(net_fd, "DOTD", &len) == 0)
is ok. it's multiline if statement.

iif ((ret = dcc_cpp_maybe(argv, input_fname, &cpp_fname, &cpp_pid) != 0))
indeed, either error or misuse/unclear usage, it might be done intentionally, need some investigation.
